### ISSUE TITLE
jax.experimental.mosaic.gpu.profiler.measure is missing a warmup?

### ISSUE THREAD

**Issue by mariogeiger** [jax-ml/jax#26144]
### Description

I'm getting a runtime too long the first time I profile.
I expect `profiler.measure` to take care of compiling and warming up.

```python
import jax.numpy as jnp
from jax.experimental.mosaic.gpu import profiler

inputs = [jnp.zeros((4000, 4000)), jnp.zeros((4000, 4000))]


def f(x, y):
    out = 0
    for i in range(0, 4000, 1000):
        out += x[i : i + 1000].T @ y[i : i + 1000]
    return out


_, runtime = profiler.measure(f, mode="events")(*inputs)
print(f"Runtime measured with events: {runtime:.2f} ms")
# Runtime measured with events: 1.73 ms

_, runtime = profiler.measure(f, mode="cupti")(*inputs)
print(f"Runtime measured with cupti: {runtime:.2f} ms")
# Runtime measured with cupti: 215.84 ms 😱😱😱

_, runtime = profiler.measure(f, mode="cupti")(*inputs)
print(f"Runtime measured with cupti: {runtime:.2f} ms")
# Runtime measured with cupti: 1.61 ms
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.5.0
jaxlib: 0.5.0
numpy:  2.2.2
python: 3.12.7 (main, Oct  1 2024, 08:52:12) [GCC 11.4.0]
device info: NVIDIA RTX 5880 Ada Generation-2, 2 local devices"
process_count: 1
platform: uname_result(system='Linux', node='db1ac34-lcedt', release='6.8.0-49-generic', version='#49~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Nov  6 17:42:15 UTC 2', machine='x86_64')


$ nvidia-smi
Tue Jan 28 02:08:14 2025       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 560.28.03              Driver Version: 560.28.03      CUDA Version: 12.6     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA RTX 5880 Ada Gene...    Off |   00000000:01:00.0 Off |                  Off |
| 30%   48C    P0             46W /  285W |     350MiB /  49140MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
|   1  NVIDIA RTX 5880 Ada Gene...    Off |   00000000:09:00.0 Off |                  Off |
| 30%   44C    P0             29W /  285W |     346MiB /  49140MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A   3601092      C   ...iger/envs/cuequivariance/bin/python        340MiB |
|    1   N/A  N/A   3601092      C   ...iger/envs/cuequivariance/bin/python        336MiB |
+-----------------------------------------------------------------------------------------+
```

**Comment by andportnoy** [comment]
Originally I skipped warmup in the CUPTI-based implementation since it isn't affected by host runtime (unlike the events-based implementation). But as you are showing here, there are cases when the measurements in the first run can be affected by the autotuner on the device:
```python
# note the aggregate=False
fprof = profiler.measure(f, mode="cupti", aggregate=False)

_, runtime_warmup = fprof(*inputs)
_, runtime        = fprof(*inputs)

print(f"{len(runtime_warmup)=}")
# len(runtime_warmup)=881
# the XLA autotuner runs many many kernels

print(f"{len(runtime)=}")
# len(runtime)=5
# only 5 are actually required for execution

print(f"{runtime=}")
# runtime=[('gemm_fusion_dot_10_0', 0.443844), ('gemm_fusion_dot_6_0', 0.443973), ('gemm_fusion_dot_20_0', 0.444996), ('gemm_fusion_dot_15_0', 0.444293), ('loop_add_fusion', 0.350275)]
```

Maybe we should change the CUPTI-based profiler to warm up just like the events-based profiler does.
